modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/IgniteSnapshotManager.java

```diff
@@ -0,0 +1,1410 @@
+                                    file::toPath,
+                                    new LongAdderMetric("NO_OP", null));
+
+                            transFut.stores.put(new GroupPartitionId(grpId, partId), pageStore);
+
+                            pageStore.init();
+                        }
+                        catch (IgniteCheckedException e) {
+                            throw new IgniteException(e);
+                        }
+                        finally {
+                            busyLock.leaveBusy();
+                        }
+                    }
+                };
+            }
+        });
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void stop0(boolean cancel) {
+        busyLock.block();
+
+        try {
+            for (SnapshotFutureTask sctx : locSnpTasks.values()) {
+                // Try stop all snapshot processing if not yet.
+                sctx.acceptException(new NodeStoppingException("Snapshot has been cancelled due to the local node " +
+                    "is stopping"));
+            }
+
+            locSnpTasks.clear();
+
+            RemoteSnapshotFuture snpTrFut = rmtSnpReq.get();
+
+            if (snpTrFut != null)
+                snpTrFut.cancel();
+
+            // Do not shutdown immediately since all task must be closed correctly.
+            snpRunner.shutdown();
+
+            cctx.kernalContext().io().removeMessageListener(DFLT_INITIAL_SNAPSHOT_TOPIC);
+            cctx.kernalContext().event().removeDiscoveryEventListener(discoLsnr);
+            cctx.kernalContext().io().removeTransmissionHandler(DFLT_INITIAL_SNAPSHOT_TOPIC);
+        }
+        finally {
+            busyLock.unblock();
+        }
+    }
+
+    /**
+     * @return Relative configured path of presistence data storage directory for the local node.
+     * Example: {@code snapshotWorkDir/db/IgniteNodeName0}
+     */
+    public static String relativeNodePath(PdsFolderSettings pcfg) {
+        return Paths.get(DB_DEFAULT_FOLDER, pcfg.folderName()).toString();
+    }
+
+    /**
+     * @param snpLsnr Snapshot listener instance.
+     */
+    public void addSnapshotListener(SnapshotListener snpLsnr) {
+        this.snpLsnr = snpLsnr;
+    }
+
+    /**
+     * @param snpName Snapshot name.
+     * @return Local snapshot directory for snapshot with given name.
+     */
+    public File snapshotLocalDir(String snpName) {
+        assert locSnpDir != null;
+
+        return new File(locSnpDir, snpName);
+    }
+
+    /**
+     * @return Node snapshot working directory.
+     */
+    public File snapshotTempDir() {
+        assert tmpWorkDir != null;
+
+        return tmpWorkDir;
+    }
+
+    /**
+     * @param parts Collection of pairs group and appropratate cache partition to be snapshotted.
+     * @param rmtNodeId The remote node to connect to.
+     * @return Snapshot name.
+     */
+    public IgniteInternalFuture<Boolean> createRemoteSnapshot(UUID rmtNodeId, Map<Integer, Set<Integer>> parts) {
```

Comment:
I think it's better to remove `SnapshotListener` at all and add some `BiConsumer<File, GroupPartitionId>` as parameter to this method, which would be triggered for each partition processed.

Reply:
Fixed.